Week 8: Implement GET Routes Assignment

ltphongssvn opened the pull request:
## Summary

This PR adds the Week 8 assignment focused on implementing GET routes with member routes and detailed data presentation.

## Changes Made  

- **Student Dashboard Route**: Added member route GET /students/:id/dashboard
- **Dashboard Controller Action**: Implemented dashboard action with efficient data loading  
- **Dashboard View**: Comprehensive student dashboard showing enrollment details
- **Request Testing**: Complete specs for dashboard route functionality

## Key Learning Objectives

- Understand Rails member routes within resourceful routing
- Implement data presentation views with associated model data
- Practice efficient database queries with includes  
- Test GET routes with complex data relationships

Ready for review and merge.

inlibra-cpa commented on app/views/admin_dashboard/index.html.erb:
```diff
@@ -1 +1,3 @@
+<!-- /app/views/admin_dashboard/index.html.erb -->
 <h1>Dashboard</h1>
+<p>Upcoming Trimester: <%= @upcoming_trimester.term %>-<%= @upcoming_trimester.year %></p>
```
Render both sections - current and upcoming here, in the view.

inlibra-cpa commented on config/routes.rb:
```diff
@@ -6,6 +6,7 @@
   resources :lessons
   resources :courses
   resources :coding_classes
+  get "dashboard" => "admin_dashboard#index", as: :admin_dashboard
```
You may consider adding resourceful routes for `courses#show` and `mentors#index/show` so path helpers work and links are simple.

```
  resources :lessons
  resources :courses, only: [:show]
  resources :coding_classes
  resources :mentors, only: [:index, :show]

  get "dashboard", to: "admin_dashboard#index", as: :admin_dashboard
```

inlibra-cpa reviewed (commented):
Hey, @ltphongssvn ,

Great job on the assignment!
Below are suggested improvements as well as  comments in the code.
So, before the merge, update `AdminDashboardController` to set `@current_trimester = Trimester.where("start_date <= ? AND end_date >= ?", Date.current, Date.current).first `and `@upcoming_trimester = Trimester.where(start_date: Date.current..6.months.from_now).order(:start_date).first`; in `app/views/admin_dashboard/index.html.erb` guard each section with if `@current_trimester / if @upcoming_trimester` and render course lists linking titles to `course_path(course)`; ensure `CoursesController#show` eager loads with `includes(:coding_class, enrollments: :student)` and the view lists enrolled student names; add `respond_to` blocks in `MentorsController` so `index/show` return JSON when requested; choose one 404 strategy and align tests with the app (either expect ActiveRecord::RecordNotFound or rescue in ApplicationController and return :not_found); and if `config/master.key` was ever pushed, rotate credentials and scrub it from history.

Please let me know if you have any questions, we may hop on huddle to discuss.

You also may consider adding the `courses_controller.rb` to load the course title with `coding_class` and enrolled students in one query to match the week's 7 material showing the list of enrolled students!

```
class CoursesController < ApplicationController
  def show
    @course = Course
      .includes(:coding_class, enrollments: :student)
      .find(params[:id])
  end
end
```
Below is the simples show.html.erb file that displays course name and students names. Also it is handling empty enrollments.
```
<h1><%= @course.coding_class.title %></h1>

<h2>Enrolled Students</h2>
<% if @course.enrollments.any? %>
  <ul>
    <% @course.enrollments.each do |enrollment| %>
      <% student = enrollment.student %>
      <% if student %>
        <li><%= "#{student.first_name} #{student.last_name}" %></li>
      <% end %>
    <% end %>
  </ul>
<% else %>
  <p>No students enrolled yet.</p>
<% end %>

<p><%= link_to "Back to Dashboard", admin_dashboard_path %></p>
```

Best,
Inara

ltphongssvn commented:
Hi Inara,

  Thank you for the detailed code review feedback! I have implemented all the requested changes:

  ✅ AdminDashboardController Updates:
  - Added @current_trimester query with proper date range checking
  - Updated @upcoming_trimester query with correct ordering and 6-month range

  ✅ Admin Dashboard View:
  - Added conditional guards for both @current_trimester and @upcoming_trimester sections
  - Course titles now link to course_path(course) as requested

  ✅ CoursesController#show Enhancement:
  - Added eager loading with includes(:coding_class, enrollments: :student)
  - Updated view to display course name and enrolled student names
  - Proper handling for empty enrollments included

  ✅ MentorsController JSON Support:
  - Added respond_to blocks for both index and show actions
  - Both actions now return JSON when requested

  ✅ 404 Handling Standardization:
  - Aligned all tests to use :not_found status consistently
  - Rails automatically converts ActiveRecord::RecordNotFound to 404

  ✅ Security (config/master.key):
  - Already addressed in previous security audit - properly excluded from version control

  All tests are now passing. The changes have been committed to the branch and are ready for re-review.

  Commit: 58c00f7 - implement code reviewer feedback from PR #9

  Please let me know if you need any clarification or if there are additional changes needed.

  Best regards,
Thanh Phong

ltphongssvn committed:
feat: add week 8 implement GET routes assignment

- Add dashboard action to StudentsController with member route
- Implement student dashboard view showing enrollments and course details
- Add member route: GET /students/:id/dashboard for student-specific dashboards
- Include comprehensive request specs testing dashboard functionality
- Demonstrate RESTful member routes and data presentation patterns
